orbit/pkg/update/execwinapi_windows.go

```diff
@@ -0,0 +1,122 @@
+//go:build windows
+
+package update
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/registry"
+)
+
+var (
+	dllMDMRegistration *windows.LazyDLL = windows.NewLazySystemDLL("mdmregistration.dll")
+
+	// RegisterDeviceWithManagement registers a device with a MDM service:
+	// https://learn.microsoft.com/en-us/windows/win32/api/mdmregistration/nf-mdmregistration-registerdevicewithmanagement
+	procRegisterDeviceWithManagement *windows.LazyProc = dllMDMRegistration.NewProc("RegisterDeviceWithManagement")
+)
+
+// Exported so that it can be used in tools/ (so that it can be built for
+// Windows and tested on a Windows machine). Otherwise not meant to be called
+// from outside this package.
+func RunMicrosoftMDMEnrollment(args MicrosoftMDMEnrollmentArgs) error {
+	installType, err := readInstallationType()
+	if err != nil {
+		return err
+	}
+	if strings.ToLower(installType) == "server" {
+		// do not enroll, it is a server
+		return errIsWindowsServer
+	}
+	return enrollHostToMDM(args)
+}
+
+func readInstallationType() (string, error) {
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows NT\CurrentVersion`, registry.QUERY_VALUE)
+	if err != nil {
+		return "", err
+	}
+	defer k.Close()
+
+	s, _, err := k.GetStringValue("InstallationType")
+	if err != nil {
+		return "", err
+	}
+	return s, nil
+}
+
+// TODO(mna): refactor to a Windows-specific package to constrain usage of
+// unsafe to that package, once https://github.com/fleetdm/fleet/pull/12387
+// lands.
+
+// Perform the host MDM enrollment process using MS-MDE protocol:
+// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-mde/5c841535-042e-489e-913c-9d783d741267
+func enrollHostToMDM(args MicrosoftMDMEnrollmentArgs) error {
+	discoveryURLPtr, err := syscall.UTF16PtrFromString(args.DiscoveryURL)
+	if err != nil {
+		return fmt.Errorf("discovery URL to UTF16 pointer: %w", err)
+	}
+
+	// we use an empty user UPN, it is not a required argument
+	userPtr, err := syscall.UTF16PtrFromString("")
+	if err != nil {
+		return fmt.Errorf("user UPN to UTF16 pointer: %w", err)
+	}
+
+	accessTok, err := generateWindowsMDMAccessTokenPayload(args)
+	if err != nil {
+		return fmt.Errorf("generate access token payload: %w", err)
+	}
+	accessTokPtr, err := syscall.UTF16PtrFromString(string(accessTok))
+	if err != nil {
+		return fmt.Errorf("access token to UTF16 pointer: %w", err)
+	}
+
+	// pre-load the DLL and pre-find the procedure, to return a more meaningful
+	// message if those steps fail and avoid a panic.
+	if err := dllMDMRegistration.Load(); err != nil {
+		return fmt.Errorf("load MDM dll: %w", err)
+	}
+	if err := procRegisterDeviceWithManagement.Find(); err != nil {
+		return fmt.Errorf("find MDM RegisterDeviceWithManagement procedure: %w", err)
+	}
+
+	if code, _, err := procRegisterDeviceWithManagement.Call(
```

Comment:
nit, I think it would be useful to add a debug log line with return code and err. This could help to triage field issues where RegisterDeviceWithManagement might not behave as expect4ed

Reply:
@marcosd4h any error returned here will get logged by the config fetcher middleware here: https://github.com/fleetdm/fleet/pull/12426/files#diff-b5459f59793c455e03b32f0ae9f6a52334f0fc85dcb14ba4402318af7348e77fR153

Did you have something else in mind?

Reply:
yes, I was mostly thinking about the scenario where `code == windows.ERROR_SUCCESS` but `err != nil`. This should theoretically not happen, but I noticed that `err` was not checked and thought about having a debug log in case unexpected things happens on the field

Reply:
Ah I see, but the error is _always_ non-`nil`, so in the case of `windows.ERROR_SUCCESS` it would always be "The operation completed successfully" or something like that. From the documentation of `windows.LazyProc`:

> The returned error is always non-nil, constructed from the result of GetLastError. Callers must inspect the primary return value to decide whether an error occurred

https://pkg.go.dev/golang.org/x/sys/windows#LazyProc.Call

Reply:
Debug log added.